### PR TITLE
Working synb0 wrapper in scilpy

### PR DESCRIPTION
**Pull request by frheault**
# Quick description

Please include a summary of the changes and the related issue(s) or improvement(s).
Please also include relevant motivation and context. List any dependencies that are required for this change if needed.

By providing a skull strip t1w and b0, this generate a perfect b0 without distortion which can be fed to topup (saying it has infinite bandwitdh) to trick topup into fixing the diffusion without a rev_b0.

Useful when no rev_b0 is available. Could help to simplify Tractoflow?

https://pubmed.ncbi.nlm.nih.gov/31075422/

...

## Type of change

Check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Provide data, screenshots, command line to test (if relevant)

...

# Checklist

- [ ] My code follows the style guidelines of this project (run [autopep8](https://pypi.org/project/autopep8/))
- [ ] I added relevant citations to scripts, modules and functions docstrings and descriptions
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I moved all functions from the script file (except the argparser and main) to scilpy modules
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-734/2/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-734/3/

**Comment by frheault** [comment]
The test requires tensorflow, should I just delete the test or we turn it off?

**Comment by AlexVCaron** [comment]
Could the test run conditionally to having Tensorflow installed ? https://stackoverflow.com/questions/36377625/how-to-conditionally-skip-a-test-in-python

Maybe the import error is too early and would prevent this to work ...

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_prepare_topup_command.py [hunk 1]
```diff
@@ -119,9 +122,9 @@ def main():
     fused_b0s_path = os.path.join(args.out_directory, args.out_b0s)
     nib.save(nib.Nifti1Image(fused_b0s, b0_img.affine), fused_b0s_path)
 
-    acqparams = create_acqparams(
-        args.readout, args.encoding_direction, b0.shape[-1], rev_b0.shape[-1])
-
+    acqparams = create_acqparams(args.readout, args.encoding_direction,
+                                 args.synb0, b0.shape[-1], rev_b0.shape[-1])
+    print(acqparams)
```
Comment (arnaudbore) [inline scripts/scil_prepare_topup_command.py:127]: Remove print

File: scilpy/utils/image.py [hunk 2]
```diff
@@ -104,17 +106,17 @@ def transform_dwi(reg_obj, static, dwi, interpolation='linear'):
 
 
 def register_image(static, static_grid2world, moving, moving_grid2world,
-                   transformation_type='affine', dwi=None):
+                   transformation_type='affine', dwi=None, fine=False):
```
Comment (arnaudbore) [inline scilpy/utils/image.py:109]: we need some docstring here

**Review by arnaudbore — COMMENTED** [review]
Need some testing before moving forward. Overall synthesize_b0 works well but still needs to be tested with topup.

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-734/9/

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_synthesize_b0.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                             assert_outputs_exist)
+from scilpy.utils.image import register_image
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawTextHelpFormatter)
+    p.add_argument('in_b0',
+                   help='Input b0 image.')
```
Comment (mdesco) [inline scripts/scil_synthesize_b0.py:52]: Skull-Stripped or not? You mention above that the b0 must skull-strip a few times... Please clarify... 

Idem for the T1 below

**Review by mdesco — COMMENTED** [review]
A few minor comments. Hard for me to test on my mac. Don't want to fight with tensorflow...

**Review by mdesco — APPROVED** [review]
Once the minor comments are done. All good

File: scripts/scil_synthesize_b0.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+Requires Skull-Strip b0 and t1w images as input, the script will normalize the
+t1w's WM to 110, co-register both images, then register it to the appropriate
+template, run SyNb0 and then transform the result back to the original space.
+
+This script must be used carefully, as it is not meant to be used in an
+environment with the following dependencies already installed (not default
+in Scilpy):
+- tensorflow-addons
+- tensorrt
+- tensorflow
```
[the remark this answers is not in the document]
Comment (frheault) [reply scripts/scil_synthesize_b0.py:15]: Dipy does not say either.

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by smeisler** [comment]
Thanks for your work on this, I have started testing it out.

It doesn't seem like the synb0 template fetcher function is defined in the PR, but I presume the template was the file in `/data/mni_icbm152_t1_tal_nlin_asym_09c_masked_2_5.nii.gz`? If so, the template looked strange after I visualized it when I downloaded it from GitHub:
image

The resulting b0 map also looked odd, but likely due to trying to register to that odd template image:
image

**Comment by smeisler** [comment]
Perhaps you should put this file instead in /data: https://github.com/MASILab/Synb0-DISCO/blob/master/atlases/mni_icbm152_t1_tal_nlin_asym_09c_mask.nii.gz

**Comment by smeisler** [comment]
The template here looks better: https://github.com/MASILab/Synb0-DISCO/blob/master/atlases/mni_icbm152_t1_tal_nlin_asym_09c_mask_2_5.nii.gz

**Comment by smeisler** [comment]
Fixing the template did the trick!
image

**Comment by frheault** [comment]
> The template here looks better: https://github.com/MASILab/Synb0-DISCO/blob/master/atlases/mni_icbm152_t1_tal_nlin_asym_09c_mask_2_5.nii.gz

I will fix the function and make sure the data inside the repo is reuploaded correctly, thank you !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scripts/tests/test_synthesize_b0.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+from scilpy.io.fetcher import fetch_data, get_home, get_testing_files_dict
```
Comment (arnaudbore) [inline scripts/tests/test_synthesize_b0.py:4]: Rename this file and edit this file so it actually tests: `scil_volume_b0_synthesis.py`

File: scilpy/image/volume_operations.py [hunk 6]
```diff
@@ -1,6 +1,8 @@
 # -*- coding: utf-8 -*-
 
+import inspect
```
Comment (arnaudbore) [inline scilpy/image/volume_operations.py:3]: inspect and os does not seem to be used here

**Review by arnaudbore — CHANGES_REQUESTED** [review]
A couple of comments, almost GTG

**Pull request merged**